Add `AsyncSeq` vs `TaskSeq` comparison readme (#131)
fsprojects/FSharp.Control.TaskSeq: README.md

}
```

## Choosing between `AsyncSeq` and `TaskSeq`

> The [`AsyncSeq`][11] and `TaskSeq` library both operate on asynchronous sequences, but there are a few fundamental differences, most notably that the former _does not_ implement `IAsyncEnumerable<'T>`, but has its own same-named, but differently behaving type. Another core difference is that `TaskSeq` uses `ValueTasks` for the asynchronous computations, and `AsyncSeq` uses F#'s `Async<'T>`.

Reviewer: ```suggestion
The [`AsyncSeq`][11] and `TaskSeq` libraries both operate on asynchronous sequences, but there are a few fundamental differences. The most notable being that the former _does not_ implement `IAsyncEnumerable<'T>`, though it does have a type of that name with different semantics (not surprising; it predates the definition of the modern one). Another key difference is that `TaskSeq` uses `ValueTask`s for the asynchronous computations, whereas `AsyncSeq` uses F#'s `Async<'T>`.
```

Author: Thanks! Done.

There are more differences:

|                            | `TaskSeq`                                                                             | `AsyncSeq`                                                           |
|----------------------------|---------------------------------------------------------------------------------------|----------------------------------------------------------------------|
| **Frameworks**             | .NET 5.0+, NetStandard 2.1                                                        | .NET 5.0+, NetStandard 2.0 and 2.1, .NET Framework 4.6.1+            |

> | **Underlying type**        | `System.Collections.Generic.IAsyncEnumerable<'T>`                                 | Its own type, also called `IAsyncEnumerable<'T>`, but not compatible |

Author: I'll keep both, I used your updated text.

| **Implementation**         | State machine (statically compiled)                                               | No state machine, continuation style                                 |
| **Semantics**              | `seq`-like: on-demand                                                             | `seq`-like: on-demand                                                |

> | **Support `let!`**         | All `task`-like: `Async<'T>`, `ValueTask<'T>`, `Task<'T>` or any `GetAwaiter()`   | `Async<'T>` only                                                     |

Reviewer: ```suggestion
| **Support `let!`**         | All `task`-like: `Async<'T>`, `Task<'T>`, `ValueTask<'T>` or any `GetAwaiter()`   | `Async<'T>` only                                                     |
```
(VT last to align with next line)

Author: Tx, sharp eye!

> | **Support `do!`**          | `Async<unit>`, `Task<unit>` and `Task`, `ValueTask<unit>` and `ValueTask`         | `Async<unit>` only                                                   |

Reviewer: ```suggestion
| **Support `do!`**          | `Async<unit>`, `Task<unit>`, `Task`, `ValueTask<unit>`, `ValueTask`         | `Async<unit>` only                                                   |
```

Author: I like to keep this one, as it is deliberate (I had commas before but didn't like it), it signifies they're interchangeable semantically.

| **Support `yield!`**       | `IAsyncEnumerable<'T>`, `AsyncSeq`, any sequence                                  | `AsyncSeq`                                                           |

> | **Support `for`**          | `IAsyncEnumerable<'T>`, `AsyncSeq`, any sequence                                  | `AsyncSeq` any sequence                                              |

Reviewer: ```suggestion
| **Support `for`**          | `IAsyncEnumerable<'T>`, `AsyncSeq`, any sequence                                  | `AsyncSeq`, 
 any sequence                                              |
```

> | **Behavior with `yield`**  | Zero allocations, no `Task` or even `ValueTask` created                           | Allocated, an F# `Async` wrapped in a singleton `AsyncSeq`           |

Reviewer: ```suggestion
| **Behavior with `yield`**  | Zero allocations; no `Task` or even `ValueTask` created                           | Allocates an F# `Async` wrapped in a singleton `AsyncSeq`           |
```

Author: Thanks!

| **Conversion to other**    | `TaskSeq.toAsyncSeq`                                                              | `AsyncSeq.toAsyncEnum`                                               |

> | **Conversion from other**  | Implicit (yield!) or `TaskSeq.ofAsyncSeq`                                         | `AsyncSeq.ofAsyncEnum`                                               |

Reviewer: ```suggestion
| **Conversion from other**  | Implicit (`yield!`) or `TaskSeq.ofAsyncSeq`                                         | `AsyncSeq.ofAsyncEnum`                                               |
```

> | **Recursion in `yield!`**  | No (requires F# support, upcoming)                                                | Yes                                                                  |

Reviewer: ```suggestion
| **Recursion in `yield!`**  | **NO** (requires F# support, upcoming)                                                | Yes                                                                  |
```
capital O can be dropped, but good to make people think...

Author: Good point

> | **Based on F# concept of** | `task`                                                                            | `async`                                                              |

Reviewer: If this is going to be mentioned, it needs to be on line 2 or 3?

Author: Let's keep it separate for now.

| **`MoveNextAsync`**        | `ValueTask<bool>`                                                                 | `Async<'T option>`                                                   |

> | **`Current` internals**    | `ValueOption<'T>`                                                                 | `Option<'T>`                                                         |

Reviewer 2: I would group these 2 in a single MoveNext/Current row. I think it's clearer as in AsyncSeq both things are in a single type.

Author: Tx. I'll probably just drop the second line altogether. Primarily, I don't want the left-most column to be too wordy and I can't think of a good way to explain both. Curious people can always look it up.

Author: Done, tx.

> | **Cancellation**           | Implicit token governing iteration                                                | Implicit token passed to each subtask                                |

Reviewer: ```suggestion
| **Cancellation**           | Implicit token governs iteration                                                | Implicit token flows to all subtasks per normal `Async` semantics                               |
```
Is "Implicit token governs iteration" -> "Implicit token governs iteration but requires explicit propagation of token to child Tasks"

Author: applied, tx

| **Performance**            | Very high, negligible allocations                                                 | Slower, more allocations, due to using `async`                       |

> | **Parallelism**            | Possible with ChildTask, support will follow                                      | Supported explicitly                                                 |

Reviewer: ```suggestion
| **Parallelism**            | Possible with ChildTask; support will follow                                      | Extensive explicit helpers provided                                                 |
```

Author: I'll take the semicolon, keep the original wording. People can always go to the docs to find out.

## Status & planning

> This project has stable features currently, but before we go full "version one", we'd like to complete the surface area. This section covers the status of that, with a full list of implmented functions below. Here's the short list:

Reviewer: ```suggestion
This project has stable features currently, but before we go full "version one", we'd like to complete the surface area. This section covers the status of that, with a full list of implemented functions below. Here's the shortlist:
```

Author: Oops! lol. Tx.
